### PR TITLE
Fadi/test 1610899

### PR DESCRIPTION
**Pull request by fadidurah**
This PR includes an added test in ADAL to allow removal of this manual E2E test case by replacing it with a component test https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1610899.

Note that the component test is not meant to cover this entire E2E test case, just the adal client side request/response handling, as the Broker behavior for this flow is already tested in msal UI automation.

### CODE REVIEW

File: adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java [hunk 1]
```diff
@@ -583,6 +590,46 @@ public void testPrompt() throws InterruptedException {
...
+        Mockito.when(mockedPackageManager.getPackageInfo(Mockito.anyString(), Mockito.anyInt())).thenReturn(mockPackageInfo);
+        mockContext.setMockedPackageManager(mockedPackageManager);
+
+        // Call acquire token which will try refresh token based on cache
+        context.acquireToken(testActivity.getTestActivity(), "resource", "clientid", "redirectUri",
+                TEST_IDTOKEN_UPN, PromptBehavior.Auto, null, "{\"access_token\":{\"deviceid\":{\"essential\":true}}}", callback);
+        signal.await(CONTEXT_REQUEST_TIME_OUT, TimeUnit.MILLISECONDS);
+
+        final JWT decodedToken = JWTParser.parse(callback.getAuthenticationResult().getAccessToken());
+        Assert.assertNotNull(decodedToken.getJWTClaimsSet().getClaim("deviceid"));
```
Comment (shahzaibj) [inline adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java]: Let's also add an assert for the value of the deviceid.

File: adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java [hunk 2]
```diff
@@ -583,6 +590,50 @@ public void testPrompt() throws InterruptedException {
...
+        mockContext.setMockedPackageManager(mockedPackageManager);
+
+        // Call acquire token which will try refresh token based on cache
+        context.acquireToken(testActivity.getTestActivity(), "resource", "clientid", "redirectUri",
+                TEST_IDTOKEN_UPN, PromptBehavior.Auto, null, "{\"access_token\":{\"deviceid\":{\"essential\":true}}}", callback);
+        signal.await(CONTEXT_REQUEST_TIME_OUT, TimeUnit.MILLISECONDS);
+
+        final JWT decodedToken = JWTParser.parse(callback.getAuthenticationResult().getAccessToken());
+        final Object deviceIdClaim = decodedToken.getJWTClaimsSet().getClaim("deviceid");
+        Assert.assertNotNull(deviceIdClaim);
```
Comment (iamgusain) [inline adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java:634]: nit: we can compare the full token with "TEST_ACCESS_TOKEN_WITH_DEVICE_ID "

File: adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java [hunk 3]
```diff
@@ -583,6 +590,50 @@ public void testPrompt() throws InterruptedException {
...
+
+        final PackageManager mockedPackageManager = Mockito.mock(PackageManager.class);
+        final PackageInfo mockPackageInfo = new PackageInfo();
+        mockPackageInfo.setLongVersionCode(Long.MAX_VALUE);
+        Mockito.when(mockedPackageManager.getPackageInfo(Mockito.anyString(), Mockito.anyInt())).thenReturn(mockPackageInfo);
+        mockContext.setMockedPackageManager(mockedPackageManager);
+
+        // Call acquire token which will try refresh token based on cache
+        context.acquireToken(testActivity.getTestActivity(), "resource", "clientid", "redirectUri",
+                TEST_IDTOKEN_UPN, PromptBehavior.Auto, null, "{\"access_token\":{\"deviceid\":{\"essential\":true}}}", callback);
```
Comment (iamgusain) [inline adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java:626]: does it matter what claim is passed, based on the mocking it is going to return the same token right, irrespective of what claims are passed.

**Review by iamgusain — APPROVED** [review]
LGTM

File: adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java (hunk 3 above)
Comment (fadidurah) [reply adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java:626]: Yes, the same token will be returned. This is moreso checking that we can pass in the claim and actually complete an acquiretoken regardless of what the result we actually get is.

File: adal/src/androidTest/java/com/microsoft/aad/adal/MockDeviceCertProxy.java [hunk 4]
```diff
@@ -77,12 +72,22 @@ public String getThumbPrint() {
         return sThumbPrint;
     }
 
+    @Override
+    public byte[] sign(String algorithm, byte[] dataToBeSigned) throws ClientException {
```
Comment (shahzaibj) [inline adal/src/androidTest/java/com/microsoft/aad/adal/MockDeviceCertProxy.java:76]: I'm wondering why this change is being needed now

**Review by shahzaibj — APPROVED** [review]

File: adal/src/androidTest/java/com/microsoft/aad/adal/MockDeviceCertProxy.java (hunk 4 above)
Comment (fadidurah) [reply adal/src/androidTest/java/com/microsoft/aad/adal/MockDeviceCertProxy.java:76]: Looks like the super class was updated in common, but this class that extends it was not updated. I was getting a complitation error.
